nextcord/http.py

```diff
@@ -124,25 +140,329 @@ def bucket(self) -> str:
+        self._first_update: bool = True
+        """If the next update to be ran will be the first."""
+        self._reset_remaining_task: Optional[asyncio.Task] = None
+        """Holds the task object for resetting the remaining count."""
+        self._on_reset_event: asyncio.Event = asyncio.Event()
+        """Used to indicate when the rate limit is ready to be acquired."""
+        self._migrating: Optional[str] = None
+        """When this RateLimit is being deprecated and acquiring requests need to migrate to a different RateLimit, this
+        variable should be set to the different RateLimit/buckets string name.
+        """
+
+        self._on_reset_event.set()
+
+    @property
+    def resetting(self) -> bool:
+        """Returns True when there is a reset task and it's active. False if there isn't a task or if it's done."""
+        return self._reset_remaining_task is not None and not self._reset_remaining_task.done()
+
+    async def update(self, response: aiohttp.ClientResponse) -> None:
+        """Updates the rate limit with information from the response."""
+
+        if response.headers.get("X-RateLimit-Global") == "true":
+            # The response is intended for the global rate limit, not a regular rate limit.
+            return
+
+        # Updates the bucket name. The bucket name not existing as fine, as ``None`` is desired for that.
+        # This is done immediately, so we can error out if we get an update not for this bucket.
+        x_bucket = response.headers.get("X-RateLimit-Bucket")
+
+        if self.bucket == x_bucket:
+            pass  # Don't need to set it again.
+        elif self.bucket is None:
+            self.bucket = x_bucket
+        else:
+            raise IncorrectBucket(
+                f"Update given for bucket {x_bucket}, but this RateLimit is for bucket {self.bucket}!"
+            )
+
+        # Updates the limit if it exists.
+        x_limit = response.headers.get("X-RateLimit-Limit")
+        self.limit = 1 if x_limit is None else int(x_limit)
+
+        # Updates the remaining left if it exists, being pessimistic.
+        x_remaining = response.headers.get("X-RateLimit-Remaining")
+
+        if x_remaining is None:
+            self.remaining = 1
+        elif self._first_update:
+            self.remaining = int(x_remaining)
+        else:
+            # If requests come back out of order, it's possible that we could get a wrong amount remaining.
+            # It's best to be pessimistic and assume it cannot go back up unless the reset task occurs.
+            self.remaining = (
+                int(x_remaining) if int(x_remaining) < self.remaining else self.remaining
+            )
+
+        # Updates the datetime of the reset.
+        x_reset = response.headers.get("X-RateLimit-Reset")
+        if x_reset is not None:
+            x_reset = datetime.fromtimestamp(float(x_reset) + self._time_offset, tz=timezone.utc)
```

Comment:
time.time is always in UTC which is a lot more readable than using datetimes and getting to weird dates close to epoch.